### ISSUE TITLE
Allow passing SRPM as URL

### ISSUE THREAD

**Issue by lachmanfrantisek** [openscanhub/openscanhub#248]
To ease the Packit integration, it would be nice to allow passing SRPM in the form of a URL. (Packit uses Copr for SRPM builds so we would need to download the file upload to OSH.)

Thanks in advance!

(Also mentioned by Tomáš in #211 )

**Comment by siteshwar** [comment]
> (Also mentioned by Tomáš in #211 )

As I understand, this comment was about running scan on a locally created SRPM and not about specifying a remote URL.

**Comment by siteshwar** [comment]
One way to implement this functionality is to download the SRPM [directly](https://github.com/openscanhub/openscanhub/blob/bd8475834908a317f552cb9321e24b2da46721b1/osh/worker/tasks/task_build.py#L84) in the workers. But in that case the srpm would not be archived by hub. We should discuss if it should be necessary to archive the source rpm by hub.

**Comment by kdudka** [comment]
@siteshwar What is the point of archiving scan results of something that you have no source code for anymore?

OSH either scans a tagged Koji build, where it expects Koji to keep the source code available, or an uploaded SRPM/tarball.  In the latter case, the uploaded source code is archived/deleted along with the scan results.

**Comment by lachmanfrantisek** [comment]
@siteshwar @kdudka Do you need the SRPM archiving for some special reason? When specifying SRPM as a URL, it should be archived in the service hosting this URL (on Copr in our case and deleted after some time).

**Comment by kdudka** [comment]
@lachmanfrantisek I do not think we have any legal obligations to do so.  I am just looking at it from user's perspective.  If a user looks at OSH results too late, they may discover something useful in the results, yet not be able to map the findings to the corresponding source code because the SRPM has already been deleted by COPR.

Given the fact that the newly introduced interface in OSH will not by used solely by COPR, it may even happen that the SRPM available through the original URL will be modified later, which would make me really upset being an OSH user.

@siteshwar If we want to get the SRPM archived by OSH (although I am not insisting on that), this can be initially implemented in `osh-cli` (and eventually optimized later on without bothering end users):
1. fetch the SRPM specified by URL to a temporary directory
2. upload the SRPM as usually
3. remove the temporary directory

@lzaoral What is your say on this?  Should OSH provide scan results for source code that cannot be accessed anymore?

**Comment by siteshwar** [comment]
> @siteshwar If we want to get the SRPM archived by OSH (although I am not insisting on that), this can be initially implemented in `osh-cli` (and eventually optimized later on without bothering end users):
> 
> 1. fetch the SRPM specified by URL to a temporary directory
> 2. upload the SRPM as usually
> 3. remove the temporary directory

As I recall, the original concern from the Packit team was they wanted to avoid the task to download the SRPM on the client side. @lachmanfrantisek Is that correct?

**Comment by lachmanfrantisek** [comment]
> Given the fact that the newly introduced interface in OSH will not be used solely by COPR, it may even happen that the SRPM available through the original URL will be modified later, which would make me really upset being an OSH user.

I agree this might be a concern in general but the Copr URL Packit wants to use should be unique.

I think it's ok to leave it to the user who provides such a URL if the behaviour is clearly documented.

**Comment by siteshwar** [comment]
> I agree this might be a concern in general but the Copr URL Packit wants to use should be unique.
> 
> I think it's ok to leave it to the user who provides such a URL if the behaviour is clearly documented.

As I understand, there are 2 ways to approach this situation:

- Download the SRPM on client side, as suggested by @kdudka and archive it. This does not resolve the original concern from the Packit team to avoid downloading and uploading the SRPM.
- Introduce a new option to specify copr build id. So the SRPM may be only downloaded on the osh worker and Copr should be responsible for archiving the SRPM.

I agree with @kdudka that authenticity of the results may suffer, if the SRPM is changed on a third-party server.

**Comment by lachmanfrantisek** [comment]
@siteshwar We can also start with the first approach (client side) and implement something more clever later without changing packit or anything outside of the osh.

**Comment by siteshwar** [comment]
As of now, the Packit team has implemented downloading SRPMs in their code.
